src-manual/Vulkan/Requirements.hs

```diff
@@ -184,162 +237,246 @@ data Requirements where
+addRequirement (RequireDeviceExtension {..}, isOptional) =
+  requireDeviceExtension isOptional deviceExtensionLayerName deviceExtensionName deviceExtensionMinVersion
+addRequirement (RequireInstanceSetting {..}, isOptional) =
+  requireInstanceSetting isOptional settingName setInstanceSetting
+addRequirement (RequireFeature {..}, isOptional) =
+  requireFeature isOptional featureName checkFeature enableFeature
+addRequirement (RequireProperty {..}, isOptional) =
+  requireProperty isOptional propertyName checkProperty
 
 
-requireVersion :: Word32 -> Requirements -> Requirements
-requireVersion ver1 req@(Requirements{version = ver2}) =
-  req {version = max ver1 ver2}
+requireVersion :: Bool -> Word32 -> Requirements -> Requirements
+requireVersion isOptional ver req@(Requirements{version = ReqAndOpt reqVer optVer})
+  | isOptional
+  = req {version = ReqAndOpt reqVer (max ver optVer)}
+  | otherwise
+  = req {version = ReqAndOpt (max ver reqVer) optVer}
 
-requireInstanceLayer :: ByteString -> Word32-> Requirements -> Requirements
-requireInstanceLayer layName layVer req@(Requirements{instanceLayers = lays}) =
-  req {instanceLayers = HashMap.insertWith max layName layVer lays}
+requireInstanceLayer :: Bool -> ByteString -> Word32-> Requirements -> Requirements
+requireInstanceLayer isOptional layName layVer req@(Requirements{instanceLayers = ReqAndOpt reqLays optLays})
+  | isOptional
+  = req {instanceLayers = ReqAndOpt reqLays (HashMap.insertWith max layName layVer optLays)}
+  | otherwise
+  = req {instanceLayers = ReqAndOpt (HashMap.insertWith max layName layVer reqLays) optLays}
 
-requireInstanceExtension :: Maybe ByteString -> ByteString -> Word32 -> Requirements -> Requirements
-requireInstanceExtension mbLayName extName extVer req@(Requirements{instanceExtensions = exts}) =
-  (case mbLayName of { Just layName -> requireInstanceLayer layName 0; Nothing -> id }) $
-    req {instanceExtensions = HashMap.insertWith (HashMap.unionWith max) mbLayName (HashMap.singleton extName extVer) exts}
+requireInstanceExtension :: Bool -> Maybe ByteString -> ByteString -> Word32 -> Requirements -> Requirements
+requireInstanceExtension isOptional mbLayName extName extVer req@(Requirements{instanceExtensions = exts})
+  | isOptional
+  = (case mbLayName of { Just layName -> requireInstanceLayer isOptional layName 0; Nothing -> id }) $
+      req
+        { instanceExtensions =
+          HashMap.insertWith (liftA2 $ HashMap.unionWith max)
+            mbLayName
+            (ReqAndOpt (HashMap.singleton extName extVer) HashMap.empty)
+            exts
+        }
+  | otherwise
+  = (case mbLayName of { Just layName -> requireInstanceLayer isOptional layName 0; Nothing -> id }) $
+      req
+        { instanceExtensions =
+          HashMap.insertWith (liftA2 $ HashMap.unionWith max)
+            mbLayName
+            (ReqAndOpt HashMap.empty (HashMap.singleton extName extVer))
+            exts
+        }
```

Comment:
I'll think about it.

Reply:
The problem is that we have a dependent type, e.g.

```haskell
props' = case has @props @struct of { Just _ -> props; _ -> struct : props }
```
So I end up duplicating all of the code that depends on `props'` until I get to the constructor which wraps up the existential so that the type becomes non-dependent again.

Reply:
I think I'll leave it as is, otherwise we have to introduce a dependent sum type that captures exactly the constraints we need, seems like it'll just make the code more confusing.